subprojects/launcher/src/test/groovy/org/gradle/launcher/daemon/configuration/BuildProcessTest.groovy

```diff
@@ -84,26 +126,99 @@ public class BuildProcessTest extends Specification {
         System.getProperty('baz') != null
     }
 
-    def "when required opts contain an immutable default setting ignore it"() {
+    def "defaults in required vm args are ignored"() {
         //if the user does not configure any jvm args Gradle uses some defaults
         //however, we don't want those defaults to influence the decision whether to use existing process or not
-        //e.g. those defaults should only be used for launching a new process
-        //TODO SF this is a bit messy, let's try to clean this up
+        given:
+        BuildProcess buildProcess = new BuildProcess(currentJvm, currentJvmOptions)
+
+        when:
+        def parametersWithDefaults = buildParameters((Iterable) null)
+        parametersWithDefaults.applyDefaultsFor(JavaVersion.VERSION_1_7)
+
+        then:
+        parametersWithDefaults.getEffectiveJvmArgs().containsAll(DaemonParameters.DEFAULT_JVM_ARGS)
+        buildProcess.configureForBuild(parametersWithDefaults)
+    }
+
+    def "user can explicitly disable default daemon args by setting jvm args to empty"() {
+        given:
+        BuildProcess buildProcess = new BuildProcess(currentJvm, currentJvmOptions)
+
+        def emptyBuildParameters = buildParameters([])
+        emptyBuildParameters.applyDefaultsFor(JavaVersion.VERSION_1_7)
+
+        when:
+        buildProcess.configureForBuild(emptyBuildParameters)
+
+        then:
+        !emptyBuildParameters.getEffectiveJvmArgs().containsAll(DaemonParameters.DEFAULT_JVM_ARGS)
+    }
+
+    def "user-defined vm args that correspond to defaults are not ignored"() {
```

Comment:
Maybe the test could have a better title.

It tests that arguments defined by the user which, by chance, correspond to the daemon default arguments, influence the decision making. In this test they do not match (because the current VM does not have any arguments).

If the same arguments would be set as daemon defaults (`applyDefaultsFor()`), the current VM would be reused. Because daemon default arguments should not influence the decision making. See test `defaults in required vm args are ignored`